rules/cache/cache.go

```diff
@@ -0,0 +1,378 @@
+
+type results struct {
+	elems  elemMap
+	source rules.Source
+}
+
+func newResults(source rules.Source) results {
+	return results{elems: make(elemMap), source: source}
+}
+
+// cache is an LRU-based read-through cache.
+type cache struct {
+	sync.RWMutex
+
+	capacity          int
+	nowFn             clock.NowFn
+	sleepFn           sleepFn
+	freshDuration     time.Duration
+	stutterDuration   time.Duration
+	evictionBatchSize int
+	deletionBatchSize int
+	invalidationMode  InvalidationMode
+
+	namespaces map[string]results
+	list       lockedList
+	evictCh    chan struct{}
+	deleteCh   chan struct{}
+	toDelete   []elemMap
+	wgWorker   sync.WaitGroup
+	closed     bool
+	closedCh   chan struct{}
+}
+
+// NewCache creates a new cache.
+func NewCache(opts Options) rules.Cache {
+	c := &cache{
+		capacity:          opts.Capacity(),
+		nowFn:             opts.ClockOptions().NowFn(),
+		sleepFn:           time.Sleep,
+		freshDuration:     opts.FreshDuration(),
+		stutterDuration:   opts.StutterDuration(),
+		evictionBatchSize: opts.EvictionBatchSize(),
+		deletionBatchSize: opts.DeletionBatchSize(),
+		invalidationMode:  opts.InvalidationMode(),
+		namespaces:        make(map[string]results),
+		evictCh:           make(chan struct{}, 1),
+		deleteCh:          make(chan struct{}, 1),
+		closedCh:          make(chan struct{}),
+	}
+
+	c.wgWorker.Add(numOngoingTasks)
+	go c.evict()
+	go c.delete()
+
+	return c
+}
+
+func (c *cache) Get(namespace string, id []byte) mrules.MatchResult {
+	c.RLock()
+	res, found := c.tryGetWithLock(namespace, id, dontSetIfNotFound)
+	c.RUnlock()
+	if found {
+		return res
+	}
+
+	c.Lock()
+	res, _ = c.tryGetWithLock(namespace, id, setIfNotFound)
+	c.Unlock()
+
+	return res
+}
+
+func (c *cache) Set(namespace string, source rules.Source) {
+	c.Lock()
+	results, exist := c.namespaces[namespace]
+	if !exist {
+		results = newResults(source)
+	} else {
+		results.source = source
+	}
+	c.namespaces[namespace] = results
+	c.Unlock()
+}
+
+func (c *cache) Delete(namespace string) {
+	c.Lock()
+	results, exists := c.namespaces[namespace]
+	if !exists {
+		c.Unlock()
+		return
+	}
+	delete(c.namespaces, namespace)
+	c.toDelete = append(c.toDelete, results.elems)
+	c.Unlock()
+
+	c.notifyDeletion()
+}
+
+func (c *cache) Close() error {
+	c.Lock()
+	if c.closed {
+		c.Unlock()
+		return errCacheClosed
+	}
+	c.closed = true
+	c.Unlock()
+
+	close(c.closedCh)
+	c.wgWorker.Wait()
+	return nil
+}
+
+// tryGetWithLock attempts to get the match result, returning true if a match
+// result is successfully determined and no further processing is required,
+// and false otherwise.
+func (c *cache) tryGetWithLock(
+	namespace string,
+	id []byte,
+	setType setType,
+) (mrules.MatchResult, bool) {
+	var res mrules.MatchResult
+	results, exists := c.namespaces[namespace]
+	if !exists {
+		return res, true
+	}
+	source := results.source
+	idHash := mid.HashFn(id)
+	elem, exists := results.elems[idHash]
+	if !exists {
+		if source == nil {
```

Comment:
Yeah, was originally thinking for namespaces without rules this would be useful, but seems we always create a new ruleset per namespace so that should be a non-issue. Will change.